File: clock_mvc.rb
require "observer"

class WatchModel
    include Observable
    def initialize
        @running = false
        @time = 0
        @last = 0.0
        Thread.start do
            loop do
                sleep 0.01
                if @running
                    now = Time.now.to_f
                    @time += now - @last
                    @last = now
                    changed
                    notify_observers(@time)
                end
            end
        end
    end
    
    def start_stop
        @last = Time.now.to_f
        @running = ! @running
    end

    def time
        @time
    end
end

class WatchView
    def initialize(model)
        model.add_observer(self)
    end
    
    def update(time)
        printf "%02d:%02d\n", time.to_i, (time - time.to_i)*100
        STDOUT.flush
    end
end

class WatchController
    def initialize
        @model = WatchModel.new
        @view = WatchView.new(@model)
        # system "stty cbreak -echo"

        begin
            @view.update(@model.time)
            loop do
                if STDIN.getc == ?q
                    break
                end
                @model.start_stop
            end
        ensure
            # system "stty cooked echo"
        end
    end
end

WatchController.new

File: clock_mvc.py
# type q to quit

import time
import threading


class Informer(object):

    """docstring for Informer"""

    def __init__(self):
        super(Informer, self).__init__()
        self.observers = []

    def add_observer(self, observer):
        if observer not in self.observers:
            self.observers.append(observer)

    def delete_observer(self, observer):
        if observer in self.observers:
            self.observers.remove(observer)

    def notify_observers(self):
        for observer in self.observers:
            observer.update()


class Observer(object):

    """docstring for Observer"""

    def __init__(self):
        super(Observer, self).__init__()

    def update(self):
        pass


class WatchModel(Informer):

    """docstring for WatchModel"""

    def __init__(self):
        super(WatchModel, self).__init__()
        self.running = False
        self.tick_time = 0
        self.last_time = 0

        t = threading.Thread(
            target=self.tick_timer, name='TimerThread')
        t.daemon = True
        t.start()

    def notify_observers(self, tick_time):
        for observer in self.observers:
            observer.update(tick_time)

    def tick_timer(self):
        while True:
            time.sleep(0.01)
            if self.running:
                now = time.time()
                self.tick_time += (now - self.last_time)
                self.last_time = now
                self.notify_observers(self.tick_time)

    def start_stop(self):
        self.last_time = time.time()
        self.running = not self.running


class WatchView(Observer):

    """docstring for WatchView"""

    def __init__(self, model):
        super(WatchView, self).__init__()
        model.add_observer(self)

    def update(self, tick_time):
        str_tick_time = "%02d.%02d" % (
            tick_time, (tick_time - int(tick_time)) * 100)
        print str_tick_time


class WatchController(object):

    """docstring for WatchController"""

    def __init__(self):
        super(WatchController, self).__init__()
        model = WatchModel()
        view = WatchView(model)
        view.update(model.tick_time)

        while True:
            std_in = raw_input('input something:')
            if std_in == 'q':
                break
            model.start_stop()

WatchController()
